journalalerts/journalalerts.go

```diff
@@ -0,0 +1,392 @@
+	log.Debug("New alerts")
+
+	instance = &JournalAlerts{
+		config: config, cursorStorage: cursorStorage,
+		instanceProvider: instanceProvider,
+		sender:           sender,
+	}
+
+	for _, substr := range instance.config.Filter {
+		if len(substr) == 0 {
+			log.Warning("Filter value has an empty string")
+
+			continue
+		}
+
+		tmpRegexp, err := regexp.Compile(substr)
+		if err != nil {
+			log.Errorf("Regexp compile error. Incorrect regexp: %s, error is: %s", substr, err)
+
+			continue
+		}
+
+		instance.filterRegexp = append(instance.filterRegexp, tmpRegexp)
+	}
+
+	if err = instance.setupJournal(); err != nil {
+		return nil, aoserrors.Wrap(err)
+	}
+
+	return instance, nil
+}
+
+// Close closes logging.
+func (instance *JournalAlerts) Close() {
+	log.Debug("Close alerts")
+
+	if instance.journalCancelFunction != nil {
+		instance.journalCancelFunction()
+
+		if err := instance.storeCurrentCursor(); err != nil {
+			log.Errorf("Can't store cursor: %s", err)
+		}
+
+		instance.journal.Close()
+	}
+}
+
+/***********************************************************************************************************************
+ * Private
+ **********************************************************************************************************************/
+
+func (instance *JournalAlerts) setupJournal() (err error) {
+	if instance.journal = SDJournal; instance.journal == nil {
+		if instance.journal, err = sdjournal.NewJournal(); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	for priorityLevel := 0; priorityLevel <= instance.config.SystemAlertPriority; priorityLevel++ {
+		if err = instance.journal.AddMatch(fmt.Sprintf("PRIORITY=%d", priorityLevel)); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	if err = instance.journal.AddDisjunction(); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = instance.journal.AddMatch("_SYSTEMD_UNIT=init.scope"); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if err = instance.journal.SeekTail(); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if _, err = instance.journal.Previous(); err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	cursor, err := instance.cursorStorage.GetJournalCursor()
+	if err != nil {
+		return aoserrors.Wrap(err)
+	}
+
+	if cursor != "" {
+		if err = instance.journal.SeekCursor(cursor); err != nil {
+			return aoserrors.Wrap(err)
+		}
+
+		if _, err = instance.journal.Next(); err != nil {
+			return aoserrors.Wrap(err)
+		}
+	}
+
+	go instance.handleChannels()
+
+	return nil
+}
+
+func (instance *JournalAlerts) handleChannels() {
+	result := sdjournal.SD_JOURNAL_APPEND
+	journalTicker := time.NewTicker(journalSavePeriod)
+
+	ctx, cancelFunction := context.WithCancel(context.Background())
+
+	instance.journalCancelFunction = cancelFunction
+
+	for {
+		select {
+		case <-journalTicker.C:
+			if err := instance.storeCurrentCursor(); err != nil {
+				log.Error("Can't store journal cursor: ", err)
+			}
+
+		case <-ctx.Done():
+			journalTicker.Stop()
```

Comment:
maybe in this case we should `return` ? to avoid possible goroutine leak

Reply:
fixed